### PR TITLE
fix(auth): parse hostname for mTLS and PSC endpoint certificate rotat…

### LINKED ISSUE #18147: Fix(auth): incorrect mTLS endpoint detection causes false positive cert rotations and breaks PSC endpoints
### Description
In PR #17928, a check was added to `AuthorizedSession.request` (and pre-existed in `AuthorizedHttp.urlopen`) to only trigger mTLS client certificate rotation on mTLS endpoints upon receiving a `401 Unauthorized` response:

```python
MTLS_URL_PREFIXES = ["mtls.googleapis.com", "mtls.sandbox.googleapis.com"]
use_mtls = self.is_mtls and any(prefix in url for prefix in MTLS_URL_PREFIXES)
```

**However, this implementation has two critical defects:**

1. False Positives via Substring Matching (prefix in url)
Using Python's in operator performs a substring search across the entire URL string (including path segments and query parameters). Any standard, non-mTLS URL that happens to contain mtls.googleapis.com anywhere in its path or query string (e.g. https://storage.googleapis.com/my-bucket/logs/mtls.googleapis.com.csv or https://logging.googleapis.com/v2/entries?filter=mtls.googleapis.com) evaluates to True. When an OAuth access token expires on these non-mTLS URLs, the transport falsely assumes the client certificate expired, triggering unnecessary certificate parsing, disk I/O, and SSL adapter reconfiguration.

2. False Negatives on Private Service Connect (PSC) Endpoints
Enterprise customers connecting via Private Service Connect (PSC) use endpoints under *.p.googleapis.com (e.g. https://storage.p.googleapis.com or https://pubsub.p.googleapis.com). These endpoints use mTLS, but because their hostname does not contain the substring mtls.googleapis.com, use_mtls evaluates to False. As a result, client certificate rotation upon certificate expiration is permanently broken for PSC users.

**Expected Behavior**

- Only genuine mTLS endpoints (e.g. *.mtls.googleapis.com, *.mtls.sandbox.googleapis.com, and *.p.googleapis.com) should trigger certificate rotation checks.
- URL paths and query parameters should not influence mTLS endpoint detection.

**Actual Behavior**

- Non-mTLS URLs containing mtls.googleapis.com in paths/queries trigger certificate rotation on 401.
- PSC mTLS endpoints (*.p.googleapis.com) fail to trigger certificate rotation on 401.

**Proposed Solution**

1. Introduce a centralized helper is_mtls_endpoint(url) in google.auth.transport._mtls_helper that isolates and parses the hostname using urllib.parse.urlsplit(url).hostname.
2. Validate the hostname against:
  - Exact hosts: mtls.googleapis.com, mtls.sandbox.googleapis.com
  - Standard mTLS suffixes: .mtls.googleapis.com, .mtls.sandbox.googleapis.com
  - Private Service Connect (PSC) suffixes: .p.googleapis.com
3. Update both requests.py (AuthorizedSession) and urllib3.py (AuthorizedHttp) to use this shared helper.

### PR DESCRIPTION
**Pull request by attharva-24**
Fixes #18147
Follow-up to #17928

### Description
This PR resolves two defects in the mTLS endpoint detection logic previously used in `requests.py` and `urllib3.py`:

1. **Eliminates False Positives**: Replaces raw URL substring search (`prefix in url`) with proper hostname isolation via `urllib.parse.urlsplit(url).hostname`. Standard non-mTLS URLs containing `mtls.googleapis.com` in paths or query parameters (e.g. `https://storage.googleapis.com/bucket/mtls.googleapis.com` or `https://logging.googleapis.com/v2/entries?filter=mtls.googleapis.com`) will no longer trigger unnecessary certificate rotation on 401.
2. **Adds Private Service Connect (PSC) Support**: Adds support for enterprise PSC custom mTLS domains (`*.p.googleapis.com`) and regional mTLS domains (`*.rep.mtls.googleapis.com`), ensuring certificate rotation functions correctly for PSC connections.
3. **Centralizes Endpoint Helper**: Adds `_mtls_helper.is_mtls_endpoint(url)` shared across both `requests` and `urllib3` transports, with lazy evaluation on 401 status codes.

### Tests
- Added `TestIsMtlsEndpoint` unit test suite in `tests/transport/test__mtls_helper.py` covering standard mTLS, PSC endpoints, regional endpoints, path/query substring traps, port numbers, and edge cases.
- Added integration tests in `tests/transport/test_requests.py` and `tests/transport/test_urllib3.py` verifying cert rotation is skipped on non-mTLS URLs with matching substrings and triggered on PSC URLs.

### CODE REVIEW

File: packages/google-auth/google/auth/transport/_mtls_helper.py [hunk 1]
```diff
@@ -840,3 +841,42 @@ def call_client_cert_callback():
...
+        return False
+    if isinstance(url, bytes):
+        try:
+            url = url.decode("utf-8")
+        except (UnicodeDecodeError, AttributeError):
+            return False
+    try:
+        hostname = urlsplit(url).hostname
+    except (ValueError, TypeError, AttributeError):
+        return False
```
[the remark this answers is not in the document]
Comment (parthea) [reply packages/google-auth/google/auth/transport/_mtls_helper.py:881]: @attharva-24 PTAL

[the remark this answers is not in the document]
Comment (attharva-24) [reply packages/google-auth/google/auth/transport/_mtls_helper.py:881]: Addressed in the latest commit! 

`is_mtls_endpoint` now checks for `.url` on `urllib3.util.Url` (and other URL objects), supports `bytes` inputs with safe UTF-8 decoding, and falls back to string conversion. Added unit tests covering `urllib3.util.Url`, `bytes`, and `str` inputs across standard, PSC, and regional mTLS endpoints.

File: packages/google-auth/google/auth/transport/_mtls_helper.py [hunk 2]
```diff
@@ -840,3 +841,46 @@ def call_client_cert_callback():
...
+        url = str(url)
+    try:
+        hostname = urlsplit(url).hostname
+    except (ValueError, TypeError, AttributeError):
+        return False
+
+    if not hostname:
+        return False
+
+    hostname = hostname.lower()
```
Comment (macastelaz) [inline packages/google-auth/google/auth/transport/_mtls_helper.py]: nit: I believe technically FQDNs can end with a trailing root dot (e.g. "https://pubsub.mtls.googleapis.com." is technically valid - although not expected in the wild). For complete coverage, perhaps we should also strip "." (e.g. hostname = hostname.rstrip(".").lower();

File: packages/google-auth/tests/transport/test__mtls_helper.py [hunk 3]
```diff
@@ -1888,3 +1889,60 @@ def test_remove_oserror_ignored(
         mock_fh.flush.assert_called_once()
         mock_fsync.assert_called_once()
         mock_remove.assert_called_once_with("/path/to/secret")
+
+
+class TestIsMtlsEndpoint(object):
+    @pytest.mark.parametrize(
+        "url",
+        [
```
Comment (macastelaz) [inline packages/google-auth/tests/transport/test__mtls_helper.py:1897]: This lacks examples with explicit port numbers (example: https://pubsub.mtls.googleapis.com:443/v1) and queries and fragments (e.g. "https://pubsub.mtls.googleapis.com/v1/projects?pageSize=10#frag")

File: packages/google-auth/tests/transport/test__mtls_helper.py [hunk 4]
```diff
@@ -1888,3 +1889,60 @@ def test_remove_oserror_ignored(
...
+            urllib3.util.parse_url("https://pubsub.mtls.googleapis.com/v1"),
+            urllib3.util.parse_url("https://storage.p.googleapis.com/b/my-bucket"),
+        ],
+    )
+    def test_is_mtls_endpoint_true(self, url):
+        assert _mtls_helper.is_mtls_endpoint(url) is True
+
+    @pytest.mark.parametrize(
+        "url",
+        [
```
Comment (macastelaz) [inline packages/google-auth/tests/transport/test__mtls_helper.py:1937]: Consider adding a bare PSC case (example "https://p.googleapis.com").

File: packages/google-auth/google/auth/transport/_mtls_helper.py [hunk 5]
```diff
@@ -840,3 +841,46 @@ def call_client_cert_callback():
...
+    ".mtls.sandbox.googleapis.com",
+    ".p.googleapis.com",
+)
+_MTLS_EXACT_HOSTS = (
+    "mtls.googleapis.com",
+    "mtls.sandbox.googleapis.com",
+)
+
+
+def is_mtls_endpoint(url: Optional[Union[str, bytes, Any]]) -> bool:
```
Comment (macastelaz) [inline packages/google-auth/google/auth/transport/_mtls_helper.py]: instead of Any can you use "object"? I think ideally we'd avoid usage of Any wherever possible.

File: packages/google-auth/tests/transport/test__mtls_helper.py (hunk 4 above)
Comment (macastelaz) [reply packages/google-auth/tests/transport/test__mtls_helper.py:1937]: Additionally a case like https://[2001:db8::1]:443/mtls.googleapis.com would be good to demonstrated handling of IPv6 syntax handling

File: packages/google-auth/tests/transport/test_requests.py [hunk 6]
```diff
@@ -937,9 +940,19 @@ def test_cert_rotation_logic_skipped_on_other_refresh_status_codes(self):
             # Assert mTLS check logic was SKIPPED (Inner Check was False)
             assert not mock_helper.check_parameters_for_unauthorized_response.called
 
-    def test_cert_rotation_skipped_on_non_mtls_url(self):
+    @pytest.mark.parametrize(
```
Comment (macastelaz) [inline packages/google-auth/tests/transport/test_requests.py]: I think this is redundant - these transports are no longer responsible for mtls checks themselves and the thing under test here shouldn't be if various forms on non-mtls endpoints are detected correctly (that is already covered in the new mtls_helper tests). Instead, I'd suggest just covering one example of mtls and one example of non mtls here to cover the requests logic specifically. Same for urllib3

File: packages/google-auth/google/auth/transport/_mtls_helper.py (hunk 2 above)
Comment (attharva-24) [reply packages/google-auth/google/auth/transport/_mtls_helper.py]: Added .rstrip(".") before matching so trailing root dots on FQDNs are stripped cleanly, along with unit test cases for standard mTLS, PSC, exact hosts, and edge cases.

6f85765bb08e509f084bcc1e925c7021d2bced42

File: packages/google-auth/tests/transport/test__mtls_helper.py (hunk 3 above)
Comment (attharva-24) [reply packages/google-auth/tests/transport/test__mtls_helper.py:1897]: Added unit test cases in `TestIsMtlsEndpoint` covering explicit port numbers (:443, :8443), query parameters, URL fragments, and combined port+query+fragment permutations for standard mTLS, PSC, and non-mTLS endpoints.

a9eb0be

File: packages/google-auth/tests/transport/test__mtls_helper.py (hunk 4 above)
Comment (attharva-24) [reply packages/google-auth/tests/transport/test__mtls_helper.py:1937]: Added "p.googleapis.com" to `_MTLS_EXACT_HOSTS` so bare apex PSC domains are recognized as mTLS endpoints, along with unit test cases for "https://p.googleapis.com", ports, and trailing root dots.

Added IPv6 test cases (https://[2001:db8::1]:443/mtls.googleapis.com and https://[::1]:8443/mtls.googleapis.com) to confirm that bracketed IPv6 host syntax is handled properly.

9dbd745

File: packages/google-auth/google/auth/transport/_mtls_helper.py (hunk 5 above)
Comment (attharva-24) [reply packages/google-auth/google/auth/transport/_mtls_helper.py]: Replaced `Any` with `object` in the function signature and docstrings, and removed the unused `Any` import.

f20b946

File: packages/google-auth/tests/transport/test_requests.py (hunk 6 above)
Comment (macastelaz) [reply packages/google-auth/tests/transport/test_requests.py]: Looks like just this comment is pending and then I'll take one more look.

Comment (attharva-24) [reply packages/google-auth/tests/transport/test_requests.py]: Simplified the transport test cases. One represents mTLS endpoint and one represents non-mTLS endpoint, while keeping the full endpoint permutations covered in `test__mtls_helper.py`.

dafe9d1

**Review by macastelaz — APPROVED** [review]
